chainntnfs/bitcoindnotify/bitcoind.go

```diff
@@ -330,50 +330,95 @@ out:
-			spenderSha := msgTx.TxHash()
-			spendDetails := &chainntnfs.SpendDetail{
-				SpentOutPoint:     &prevOut,
-				SpenderTxHash:     &spenderSha,
-				SpendingTx:        &msgTx,
-				SpenderInputIndex: uint32(i),
-			}
-			// TODO(roasbeef): after change to
-			// loadfilter, only notify on block
-			// inclusion?
-			if tx.Block != nil {
-				spendDetails.SpendingHeight = tx.Block.Height
-			} else {
-				spendDetails.SpendingHeight = bestHeight + 1
-			}
+		outpointsSpent = append(outpointsSpent, prevOut)
 
-			for _, ntfn := range clients {
-				chainntnfs.Log.Infof("Dispatching "+
-					"spend notification for "+
-					"outpoint=%v", ntfn.targetOutpoint)
-				ntfn.spendChan <- spendDetails
+		spenderSha := msgTx.TxHash()
+		spendDetails := &chainntnfs.SpendDetail{
+			SpentOutPoint:     &prevOut,
+			SpenderTxHash:     &spenderSha,
+			SpendingTx:        &msgTx,
+			SpenderInputIndex: uint32(i),
+		}
 
-				// Close spendChan to ensure that any calls to Cancel will not
-				// block. This is safe to do since the channel is buffered, and the
-				// message can still be read by the receiver.
-				close(ntfn.spendChan)
-			}
-			delete(b.spendNotifications, prevOut)
+		// TODO(roasbeef): after change to loadfilter, only notify on
+		// block inclusion?
+		if tx.Block != nil {
+			spendDetails.SpendingHeight = tx.Block.Height
+		} else {
+			spendDetails.SpendingHeight = bestHeight + 1
+		}
+
+		for _, ntfn := range clients {
+			chainntnfs.Log.Infof("Dispatching spend notification "+
+				"for outpoint=%v", ntfn.targetOutpoint)
+
+			ntfn.spendChan <- spendDetails
+
+			// Close spendChan to ensure that any calls to Cancel
+			// will not block. This is safe to do since the channel
+			// is buffered, and the message can still be read by the
+			// receiver.
+			close(ntfn.spendChan)
+		}
+
+		delete(b.spendNotifications, prevOut)
+	}
+
+	// Finally, remove the spend hint from the cache once the spend
+	// notifications have been sent.
+	if len(outpointsSpent) > 0 {
+		err := b.hintCache.PurgeSpendHint(outpointsSpent...)
+		if err != nil {
+			chainntnfs.Log.Errorf("Unable to clear spend hint for "+
+				"%v: %v", outpointsSpent, err)
 		}
 	}
 }
 
+// handleBlocksConnected applies a chain update for a new block. This will
+// dispatch notifications requested by subscribers.
+func (b *BitcoindNotifier) handleBlockConnected(block chain.BlockConnected,
+	txns []*btcutil.Tx) error {
+
+	// First we'll notify any subscribed clients of the block.
+	b.notifyBlockEpochs(int32(block.Height), &block.Hash)
+
+	// Then, we'll update the spend height hint for all of our watched
+	// outpoints. This is safe to do as we do not watch already spent
+	// outpoints for spend notifications.
+	ops := make([]wire.OutPoint, 0, len(b.spendNotifications))
+	for op := range b.spendNotifications {
+		ops = append(ops, op)
+	}
+
+	if len(ops) > 0 {
+		err := b.hintCache.CommitSpendHint(uint32(block.Height), ops...)
```

Comment:
Would it be safe to do this before notifying the epochs? If so, I think it would produce more consistent behavior since the state of db is finalized before triggering downstream actions

Reply:
Not sure it would help as there's another database transaction within `txConfNotifier.ConnectTip`.